File: CSharp/NetCubeSolver/ExactCover/ThreadedDancingLinks.cs
using System.Collections;

namespace NetCubeSolver.ExactCover;

public class ThreadedDancingLinks
{
    
    private List<int> MatrixColumnNames;
    private List<BitArray> CoverMatrix;
    private List<CoverNode> PartialSolution = new List<CoverNode>();
    private ISolver Solver;
    private Config CubeConfig;

    public ThreadedDancingLinks(List<int> matrixColumnNames, List<BitArray> coverMatrix, List<CoverNode> partialSolution,
        ISolver solver, Config cubeConfig)
    {
        MatrixColumnNames = matrixColumnNames;
        CoverMatrix = coverMatrix;
        PartialSolution = partialSolution;
        Solver = solver;
        CubeConfig = cubeConfig;
    }

    public void Search(object? stateInfo)
    {
        var dancingLinks = new DancingLinks(MatrixColumnNames, CoverMatrix);
        dancingLinks.ApplyPartialCover(PartialSolution);
        dancingLinks.Search();

        var solutions = dancingLinks.ConvertedSolutions(CubeConfig, PartialSolution);
        
        Dictionary<string, PuzzleSolution> uniqueSolutions = new();
        int duplicateSolutions = 0;

        foreach (var solution in solutions)
        {
            var canonicalString = solution.FindAndSetToCanonicalSolution();
            if (uniqueSolutions.ContainsKey(canonicalString))
            {
                duplicateSolutions++;
            }
            else
            {
                uniqueSolutions.Add(canonicalString, solution);
            }
        }
        
        Solver.AddSolutions(uniqueSolutions, duplicateSolutions);
        dancingLinks.Clear();
        PartialSolution.Clear();
    }
    
}

File: Kotlin/app/src/main/kotlin/ca/jwolf/cubesolver/dancinglinks/ThreadedDancingLinks.kt
package ca.jwolf.cubesolver.dancinglinks

import ca.jwolf.cubesolver.configs.PuzzleConfig
import ca.jwolf.cubesolver.datamodels.CoverNode
import ca.jwolf.cubesolver.datamodels.PuzzleSolution
import ca.jwolf.cubesolver.datamodels.Solver
import java.util.BitSet

class ThreadedDancingLinks(private val maxtrixColumnNames: List<Int>, private val coverMatrix: List<BitSet>,
                           private val partialSolution: List<CoverNode>, private val solver: Solver,
                           private val config: PuzzleConfig) {


    suspend fun search(){
        val dlx = DancingLinks(maxtrixColumnNames, coverMatrix)
        dlx.applyPartialCover(partialSolution)
        dlx.search()

        val solutions = dlx.getPuzzleSolutions(config, partialSolution)

        val uniqueSolutions: MutableMap<String, PuzzleSolution> = mutableMapOf()
        var duplicates = 0

        solutions.forEach { solution ->
            val canonicalString = solution.getCanonicalSolutionString()
            if (uniqueSolutions.containsKey(canonicalString)) {
                duplicates++
            } else {
                uniqueSolutions[canonicalString] = solution
            }
        }

        solver.addSolutions(uniqueSolutions, duplicates)
        dlx.clear()

    }

}